### PR TITLE
image: support prepare-image --classic for snapd snap only images

### PR DESCRIPTION
**Pull request by pedronis**
To do this, the changes start with a refactor of how we emit seed.yaml to be based on stable sorting by type, this is the same sorting used at first boot time.

Consequently:

* we need to respect the user ordering but don't need to agonize on details

* we are able to move to later the decision whether on classic we need
  to add implicitly snapd or core, this is unavoidable because it requires to
  have seen all other snaps and their requirements

* later this should also allow to correctly fetch prerequisites for
  extra snaps vs required/from the model snaps

### CODE REVIEW

**Comment by pedronis** [comment]
based on #7130 and #7134

File: image/image.go [hunk 1]
```diff
@@ -521,126 +515,31 @@ func setupSeed(tsto *ToolingStore, model *asserts.Model, opts *Options, local *l
...
-		if local.IsLocal(name) {
-			fmt.Fprintf(Stdout, "Copying %q (%s)\n", local.Path(name), name)
-		} else {
-			fmt.Fprintf(Stdout, "Fetching %s\n", name)
-		}
-
-		snapChannel, err := snapChannel(name, model, opts, local)
-		if err != nil {
+	// provide snapd
+	if seed.needsCore {
```
Comment (pedronis) [inline image/image.go]: mmh, this probably should be an error instead if model.Base() != ""

File: image/image.go [hunk 2]
```diff
@@ -671,8 +574,7 @@ func setupSeed(tsto *ToolingStore, model *asserts.Model, opts *Options, local *l
 		}
 	}
 
-	// TODO: add the refs as an assertions list of maps section to seed.yaml
-
+	seedYaml := seed.seedYaml()
```
Comment (mvo5) [inline image/image.go]: (nitpick^99) I think I would switch this line with the following line.

File: image/image.go [hunk 3]
```diff
@@ -710,6 +612,205 @@ func setupSeed(tsto *ToolingStore, model *asserts.Model, opts *Options, local *l
...
+	snap     *snap.SeedSnap
+	snapType snap.Type
+}
+
+type seedEntriesByType []seedEntry
+
+func (e seedEntriesByType) Len() int      { return len(e) }
+func (e seedEntriesByType) Swap(i, j int) { e[i], e[j] = e[j], e[i] }
+func (e seedEntriesByType) Less(i, j int) bool {
+	return e[i].snapType.SortsBefore(e[j].snapType)
```
Comment (mvo5) [inline image/image.go]: Nice one

File: image/image.go [hunk 4]
```diff
@@ -521,126 +515,35 @@ func setupSeed(tsto *ToolingStore, model *asserts.Model, opts *Options, local *l
...
-		dlOpts := &DownloadOptions{
-			TargetDir: snapSeedDir,
-			Channel:   snapChannel,
+	// provide snapd or core
+	if seed.needsCore && !seed.seen["core"] {
+		// one of the snaps requires core as base
+		// which used to be implicit, add it
+		if model.Base() != "" {
+			// TODO: later turn this into an error? for sure for UC20
+			fmt.Fprintf(Stderr, "WARNING: model has base %q but some snaps require \"core\" as base as well, for compatibility it was added implicitly, listing \"core\" explicitly is recommended\n", model.Base())
```
[the remark this answers is not in the document]
Comment (chipaca) [reply image/image.go]: I think the warning (and later the error) should say _which_ snaps need core, like for core16. That would be a lot more interesting/helpful to the user.

File: image/image.go [hunk 5]
```diff
@@ -522,126 +516,35 @@ func setupSeed(tsto *ToolingStore, model *asserts.Model, opts *Options, local *l
...
-			Channel: snapChannel,
-			File:    filepath.Base(fn),
-			DevMode: info.NeedsDevMode(),
-			Classic: needsClassic,
-			Contact: info.Contact,
-			// no assertions for this snap were put in the seed
-			Unasserted: info.SnapID == "",
-		})
+	if len(seed.needsCore16) != 0 && !seed.seen["core"] {
+		return fmt.Errorf(`cannot use %s requiring base "core16" without adding "core16" explicitly or otherwise "core"`, strutil.Quoted(seed.needsCore16))
```
Comment (chipaca) [inline image/image.go]: a nit,
```suggestion
		return fmt.Errorf(`cannot use %s requiring base "core16" without adding "core16" (or "core") explicitly`, strutil.Quoted(seed.needsCore16))
```

File: image/image.go [hunk 6]
```diff
@@ -522,126 +516,35 @@ func setupSeed(tsto *ToolingStore, model *asserts.Model, opts *Options, local *l
...
 	}
-	if len(locals) > 0 {
-		fmt.Fprintf(Stderr, "WARNING: %s were installed from local snaps disconnected from a store and cannot be refreshed subsequently!\n", strutil.Quoted(locals))
+
+	if len(seed.locals) > 0 {
+		fmt.Fprintf(Stderr, "WARNING: %s were installed from local snaps disconnected from a store and cannot be refreshed subsequently!\n", strutil.Quoted(seed.locals))
```
Comment (chipaca) [inline image/image.go]: a nit: "were" is wrong if `len(seed.locals) == 1` :-)

**Comment by pedronis** [comment]
I'm landing given it is green and will do a follow up

**Pull request merged**